### PR TITLE
Add BilayerNetwork features and demo

### PR DESCRIPTION
**Pull request by jpfairbanks**
This PR adds a module for BilayerNetworks and a demonstration of using them for code generation of a fixed Petri Net to Julia program.

### CODE REVIEW

File: Project.toml
Comment (mehalter) [inline Project.toml:12]: Is this import needed? Isn't this only used in the examples?

File: Project.toml
Comment (mehalter) [inline Project.toml:15]: Is this import needed? I don't see it used anywhere else.

**Review by mehalter — COMMENTED** [review]
Is the `bilayerconversion.ipynb` file necessary? Or is it fully captured by the `.jl` file now?

We should also try to add some more tests to keep our codecov up. This currently drops our code coverage by 25%

File: Project.toml
Comment (jpfairbanks) [reply Project.toml:12]: Oh yeah, that can be removed

File: Project.toml
Comment (jpfairbanks) [reply Project.toml:15]: Probably not

**Comment by jpfairbanks** [comment]
Yeah the ipynb is not longer needed. I developed the example in an NB and then exported it to a julia file. Definitely should write some more tests. I think the example tests all the new code, so it could probably just assert that you get approximately the same answer when you solve with the bilayer and then `evaluate/compile` as when you solve with `vectorfield`

**Comment by jpfairbanks** [comment]
@mehalter any other changes?

**Comment by bosonbaas** [comment]
I know that coverage was an issue. I can quickly write up some tests that complete the codecov if that'd help get this merged

File: src/BilayerNetworks.jl
Comment (jpfairbanks) [reply src/BilayerNetworks.jl:168]: This was a fallback to make sure it didn't fail with a key not found exception

**Comment by jpfairbanks** [comment]
@mehalter is there anything blocking this PR now that @bosonbaas wrote some tests?

File: src/BilayerNetworks.jl
Comment (bosonbaas) [reply src/BilayerNetworks.jl:168]: Gotcha. This is the last piece that's lacking coverage. Could I replace this with throwing an error?

Comment (jpfairbanks) [reply src/BilayerNetworks.jl:168]: Haha, I guess proper TDD says that if we have error handling logic, then we should test that error handling. I think it is better to test it by making a bilayer with extra vertices rather than to limit the capability of the software. This is in visualization code so a bug that produces extra vertices will manifest as extra vertices with a raw numeric label and color 4. I think that is better than an exception, but could be convinced otherwise.

**Review by mehalter — APPROVED** [review]
LGTM after the tests pass

File: src/BilayerNetworks.jl
Comment (bosonbaas) [reply src/BilayerNetworks.jl:168]: Oh, if we're able to make a bilayer network that triggers this case, that'd be the ideal! I'll take a look at that

**Comment by jpfairbanks** [comment]
Merged

**Pull request merged**
